rasa/core/policies/memoization.py

```diff
@@ -313,22 +342,58 @@ def _recall_using_delorean(self, old_states, tracker, domain) -> Optional[int]:
                 old_states = states
 
             # go back again
-            mcfly_tracker = self._back_to_the_future_again(mcfly_tracker)
+            mcfly_tracker = self._back_to_the_future(mcfly_tracker, again=True)
 
         # No match found
         logger.debug(f"Current tracker state {old_states}")
         return None
 
     def recall(
-        self,
-        states: List[Dict[Text, float]],
-        tracker: DialogueStateTracker,
-        domain: Domain,
+        self, states: List, tracker: DialogueStateTracker, domain: Domain,
     ) -> Optional[int]:
+        """Finds the action based on the given states.
+
+        Uses back to the future idea to change the past and check whether the new future
+        can be used to recall the action.
 
-        recalled = self._recall_states(states)
-        if recalled is None:
+        Args:
+            states: List of states.
+            tracker: The tracker.
+            domain: The Domain.
+
+        Returns:
+            The index of the action.
+        """
+        predicted_action_index = self._recall_states(states)
+        if predicted_action_index is None:
             # let's try a different method to recall that tracker
             return self._recall_using_delorean(states, tracker, domain)
         else:
-            return recalled
+            return predicted_action_index
+
+
+def get_max_applied_events_for_max_history(
```

Comment:
Optional: Let's make this a protected method
```suggestion
def _get_max_applied_events_for_max_history(
```